src/main/scala/edu/ie3/simona/agent/grid/ReceivedValuesStore.scala

```diff
@@ -97,22 +103,28 @@ case object ReceivedValuesStore {
         (uuid, actorRefs.toVector.map(actorRef => actorRef -> None))
     }
 
-    /* Add everything, that I expect from my sub ordinate grid agents */
-    inferiorSubGridGateToActorRef.foldLeft(assetsToReceivedPower) {
-      case (subOrdinateToReceivedPower, (gate, inferiorSubGridRef)) =>
-        val couplingNodeUuid = gate.getSuperiorNode.getUuid
+    /* Add everything, that I expect from my sub ordinate grid agents. Build distinct pairs of sending actor reference
+     * and target node */
+    inferiorSubGridGateToActorRef.toVector
+      .map { case (gate, reference) =>
+        reference -> gate.getSuperiorNode.getUuid
+      }
+      .foldLeft(assetsToReceivedPower) {
+        case (
+              subOrdinateToReceivedPower,
+              (inferiorSubGridRef, couplingNodeUuid)
+            ) =>
+          /* Check, if there is already something expected for the given coupling node and add reference to the subordinate
+           * grid agent */
+          val actorRefToMessage = subOrdinateToReceivedPower
+            .getOrElse(
+              couplingNodeUuid,
+              Vector.empty[(ActorRef, Option[ProvidePowerMessage])]
+            ) :+ (inferiorSubGridRef -> None)
```

Comment:
in this case I actually think the original implementation might be easier to follow and you spare yourself a vector conversion and mapping but this might be personal preference

Reply:
The original implementation did not account for `real == 0` and the corresponding special cases. I did adapt the code a little bit though, let me know if it is easier to understand now.